### ISSUE TITLE
01. PR3: trace_memory_consistency module の public/private contract を追加する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#432]
# 01. PR3: trace_memory_consistency module の public/private contract を追加する

## 目的

PR3 で実装する trace-level memory consistency の本体を置く module を追加し、後続 issue が使う最小 contract を固定する。

この issue では **module scaffold、slot/policy helper、batch field validation helper** までを実装する。memory bank の保存ロジック、projection、loss、train.py 接続は後続 issue で扱う。

## 前提

PR1 済みの `GlobalAnchorCoarseDataset` は、training batch に以下の top-level field を出す。

```text
file_index:              (B,)    int64
primary_key_id:          (B,)    int64
secondary_key_id:        (B,)    int64
anchor_raw_indices_view: (B,H)   int64
```

PR2 済みの config は、以下の target policy 定数と typed config を持つ。

```python
TRACE_MEMORY_TARGET_POLICY_SECONDARY_CROSS = 'secondary_cross'
TRACE_MEMORY_TARGET_POLICY_PRIMARY_ANY = 'primary_any'
TRACE_MEMORY_TARGET_POLICY_PRIMARY_CROSS = 'primary_cross'
COARSE_TRACE_MEMORY_TARGET_POLICIES = (...)
CoarseTraceMemoryConsistencyCfg
```

## 対象ファイル

新規:

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```

必要なら package import を追加:

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/__init__.py
```

ただし、既存 import contract を不用意に広げる必要がなければ `__init__.py` は変更しない。

### 1. module-level constants を追加する

primary / secondary key id は PR1 の dataset contract と合わせる。

```python
TRACE_MEMORY_PRIMARY_KEY_COUNT = 3
TRACE_MEMORY_SECONDARY_KEY_COUNT = 3
TRACE_MEMORY_NUM_SLOTS = 9
```

slot は次の式で固定する。

```text
slot_id = primary_key_id * TRACE_MEMORY_SECONDARY_KEY_COUNT + secondary_key_id
```

この slot は「どの view family 由来の過去予測か」を表す。memory key は最終的に以下になる。

```text
(file_index, raw_trace_index, slot_id)
```

### 2. slot encode/decode helper を追加する

```python
def encode_trace_memory_slot(primary_key_id: int, secondary_key_id: int) -> int:
    ...


def decode_trace_memory_slot(slot_id: int) -> tuple[int, int]:
    ...
```

validation 仕様:

```text
primary_key_id:   int, 0 <= value < 3, bool 不可
secondary_key_id: int, 0 <= value < 3, bool 不可
slot_id:          int, 0 <= value < 9, bool 不可
```

bool は Python では int subclass なので明示的に拒否する。

### 3. target_policy helper を追加する

```python
def trace_memory_slot_matches_policy(
    *,
    target_policy: str,
    current_primary_id: int,
    current_secondary_id: int,
    source_primary_id: int,
    source_secondary_id: int,
) -> bool:
    ...
```

policy の意味:

```text
secondary_cross:
  source_primary_id == current_primary_id
  source_secondary_id != current_secondary_id

primary_any:
  source_primary_id / source_secondary_id は問わない

primary_cross:
  source_primary_id != current_primary_id
```

unknown policy は `ValueError` にする。message には `target_policy` と許可値を含める。

### 4. batch field helper を追加する

後続 criterion が batch を扱うときに error message を統一するため、少なくとも以下を追加する。

```python
def require_trace_memory_batch_fields(batch: dict, *, B: int, H: int) -> None:
    ...
```

確認する field:

```text
file_index              shape (B,)
primary_key_id          shape (B,)
secondary_key_id        shape (B,)
anchor_raw_indices_view shape (B,H)
meta.trace_valid        shape (B,H)
meta.fb_idx_view        shape (B,H)  # label_guard で使う。なければ後続 loss は label_guard を無効化できないため必須にする
```

`batch['meta']` が dict でない場合も明示的に `TypeError` を出す。

field dtype は厳密に固定しすぎない。DataLoader collate 後に `torch.Tensor` であること、整数 field は integer dtype であること、`trace_valid` は bool または 0/1 integer を受け付ける程度でよい。

## `__all__`

この module で外部利用する名前を `__all__` に入れる。

この issue 時点では最低限:

```python
__all__ = [
    'TRACE_MEMORY_NUM_SLOTS',
    'TRACE_MEMORY_PRIMARY_KEY_COUNT',
    'TRACE_MEMORY_SECONDARY_KEY_COUNT',
    'decode_trace_memory_slot',
    'encode_trace_memory_slot',
    'require_trace_memory_batch_fields',
    'trace_memory_slot_matches_policy',
]
```

後続 issue で memory bank / criterion を追加するときに追記する。

## 非目標

この issue では以下を実装しない。

- memory bank storage
- EMA update
- raw/view time projection
- Gaussian target generation
- consistency loss
- `TraceMemoryConsistencyCriterion`
- `train.py` への接続
- train loop metrics logging
- site54 experiment overlay

## 受け入れ条件

- `trace_memory_consistency.py` が import できる。
- slot encode/decode が round-trip する。
- `secondary_cross`, `primary_any`, `primary_cross` の policy 判定が仕様通りになる。
- invalid id / unknown policy を明示的な例外で拒否する。
- batch field helper が不足 field と shape mismatch を検出できる。

## 確認コマンド

```bash
python -m py_compile \
  packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```

PR3 の後続 test issue 完了後は以下も通す。

```bash
pytest packages/seisai-engine/tests/test_fbpick_coarse_trace_memory_consistency.py -q
```
